plugins/fluent-bit/src/plugin.cpp

```diff
@@ -0,0 +1,402 @@
+      json_buffer.pop_back(); // Remove trailing '\n'.
+      // Mutate shared state: copy generated JSON into buffer.
+      auto* state = engine->state();
+      pthread_mutex_lock(&state->lock);
+      if (json_buffer.size() > state->buf_size - state->buf_len - 1)
+        // FIXME: resize buffer instead of dying
+        die("not enough buffer capa");
+      std::memcpy(state->buf + state->buf_len, json_buffer.data(),
+                  json_buffer.size());
+      state->buf_len += json_buffer.size();
+      state->buf[state->buf_len] = '\0';
+      pthread_mutex_unlock(&state->lock);
+      co_yield {};
+    }
+    // FIXME: we're keeping things alive just for testing.
+    while (true) {
+      co_yield {};
+      std::this_thread::sleep_for(std::chrono::milliseconds(1000));
+      TENZIR_WARN("yielding...");
+    }
+  }
+
+  auto name() const -> std::string override {
+    return "fluentbit";
+  }
+
+  auto detached() const -> bool override {
+    return true;
+  }
+
+  auto location() const -> operator_location override {
+    return operator_location::local;
+  }
+
+  auto optimize(expression const& filter, event_order order) const
+    -> optimize_result override {
+    (void)order;
+    (void)filter;
+    return do_not_optimize(*this);
+  }
+
+  friend auto inspect(auto& f, fluent_bit_operator& x) -> bool {
+    return f.apply(x.args_);
+  }
+
+private:
+  operator_args args_;
+  record config_;
+};
+
+// FIXME: Shamelessly copied from the Kafka plugin. Factor it into libtenzir.
+auto kvp_parser() {
+  using namespace parsers;
+  using namespace parser_literals;
+  using parsers::printable;
+  auto key = *(printable - '=');
+  auto value = *(printable - ',');
+  auto kvp = key >> '=' >> value;
+  return kvp % ',';
+}
+
+class plugin final : public operator_plugin<fluent_bit_operator> {
+public:
+  auto initialize(const record& config, const record& /* global_config */)
+    -> caf::error override {
+    config_ = config;
+    return caf::none;
+  }
+
+  auto signature() const -> operator_signature override {
+    return {
+      .source = true,
+      .sink = true,
+    };
+  }
+
+  auto parse_operator(parser_interface& p) const -> operator_ptr override {
+    auto args = operator_args{};
+    auto parser = argument_parser{
+      name(),
+      fmt::format("https://docs.tenzir.com/docs/connectors/{}", name())};
+    auto options = std::optional<located<std::string>>{};
+    parser.add("-X,--set", options, "<key=value>,...");
+    parser.parse(p);
+    // Translate config options.
+    if (options) {
+      std::vector<std::pair<std::string, std::string>> kvps;
+      if (!kvp_parser()(options->inner, kvps))
+        diagnostic::error("invalid list of key=value pairs")
+          .primary(options->source)
+          .throw_();
+      for (auto& [key, value] : kvps)
+        args.options.emplace(std::move(key), std::move(value));
+    }
+    // Parse the remainder: <plugin> [<key=value>...]
+    auto plugin = p.accept_shell_arg();
```

Comment:
@jachris turns out this doesn't work. Do you see what I'm after? Any ideas on how to make it work?

Reply:
`argument_parser::parse(...)` checks that it consumed all arguments. The proper solution to this is to expand the argument parser to either support what you want to do, or to at least provide a way to handle unparseable arguments. The argument parser is a bit inspired by Python's `argparse`, so this could be done similar to https://docs.python.org/3.4/library/argparse.html#partial-parsing. I wanted to extend it anyway (to support n-ary arguments and so on), but did not get around to it. You can also manually parse all your arguments for now, dropping the `argument_parser` entirely.

Reply:
Got it. I switched to manual parsing for now. This is not that painful because there aren't many options here.